Assignment 4/src/FractionalWrapper.sol

```diff
@@ -0,0 +1,318 @@
+        return type(uint256).max;
+    }
+
+    /// @notice Preview the amount of assets that would be necessary when
+    ///     minting `shares`. For this contract, this returns the same as
+    ///     `convertToAssets`.
+    /// @param shares The shares to use for the mint preview.
+    /// @return assets The assets that were required as underlying.
+    function previewMint(uint256 shares) external override view returns (uint256 assets) {
+        return convertToAssets(shares);
+    }
+
+    /// @notice Mint `shares` number of shares to the `receiver`.
+    /// @param shares The shares to mint.
+    /// @param receiver The owner of the new shares.
+    /// @return assets The assets that were required as underlying.
+    function mint(uint256 shares, address receiver) external override returns (uint256 assets) {
+        assets = convertToAssets(shares);
+        balanceOf[receiver] += shares;
+        totalSupply += shares;
+        if (token.transferFrom(msg.sender, address(this), assets)) {
+            emit Deposit(msg.sender, receiver, assets, shares);
+        } else {
+            revert DepositFailed();
+        }
+    }
+
+    /// @notice The maximal amount that can be withdrawn.
+    /// @param owner The withdrawer-to-be.
+    /// @return maxAssets The maximum amount of assets.
+    /// @dev The maximum is determined solely by the balance of the user.
+    function maxWithdraw(address owner) external override view returns (uint256 maxAssets) {
+        return convertToAssets(balanceOf[owner]);
+    }
+
+    /// @notice Preview how many shares would be minted with the given assets.
+    ///     For this contract, this function does the same as
+    ///     `convertToShares`.
+    /// @param assets How many assets would be put into the wrapper.
+    /// @return shares The shares that would be minted.
+    function previewWithdraw(uint256 assets) external override view returns (uint256 shares) {
+        return convertToShares(assets);
+    }
+
+    /// @notice Internal withdraw function. Used by `withdraw` and `redeem`
+    ///     since they are identical except for the computation of
+    ///     assets/shares.
+    /// @param assets The amount of assets to withdraw.
+    /// @param shares Shares corresponding to the assets.
+    /// @param receiver The receiver of the underlying.
+    /// @param owner The owner of the wrapper tokens.
+    /// @dev This function checks the following:
+    ///     - The caller should be the owner or have sufficient allowance.
+    ///         The allowance should be subtracted.
+    ///     - The balance of the owner in shares should be sufficient.
+    ///     - The token transfer should be successful.
+    ///     In all other cases the withdrawal should revert.
+    function _withdraw(uint256 assets, uint256 shares, address receiver, address owner) internal {
+        // First check ownership/permission:
+        if (msg.sender != owner) {
```

Comment:
If you inherit from [ERC20.sol](https://github.com/yieldprotocol/yield-utils-v2/blob/6fabde585442534b54c26f444fcc752d54f230fd/contracts/token/ERC20.sol#L98), you can replace this block for a call to `_decreaseAllowance` and a call to `_transfer`.